proto/gossip_store.go

```diff
@@ -436,7 +436,14 @@ func (s *GossipStoreImpl) convertFromBytes(buf []byte, msg interface{}) error {
 func (s *GossipStoreImpl) getLocalState() types.NodeInfoMap {
 	localCopy := make(types.NodeInfoMap)
 	for key, value := range s.nodeMap {
-		localCopy[key] = value
+		cp := *(&value)         // make a copy
+		if value.Value != nil { // also copy the inner maps
+			cp.Value = make(types.StoreMap)
+			for k2, v2 := range value.Value {
+				cp.Value[k2] = v2
```

Comment:
v2 being an interface {} makes it tricky. gossip_store.go:UpdateSelf() function is called from manager.go:startHeartBeat(). node is of type api.Node. github.com/libopenstorage/openstorage/api/api.go:Node has maps inside of it. So a perfect deep copy is still not happening here. I don't know how we want to handle that or if that is read only.
We could possibly say that the maps in Node correspond to the system resources and as long as they are just ready by copies we are good. If the user of a copy modifies, then it might be a problem.

Reply:
Not saying you should do it with this change. We could possibly have an interface that has a DeepCopy() function and then have that type instead of interface{} for these maps. That way we'll force the implementer to provide a deep copy function. Just a thought.

Reply:
this is a good suggestion 👍  - this will prevent any future users from having similar issues and will automatically identify current users of the API.